File: 274 H-Index.java
/**
 * H-index I
 * https://leetcode.com/problems/h-index/
 */
public class Solution {
    
    private java.util.Random rd = new java.util.Random();
    
    public void swap(int[] arr, int lt, int rt){
        int i = arr[lt];
        arr[lt] = arr[rt];
        arr[rt] = i;
    }
    
    public int partition(int[] arr, int lt, int rt) {
        
        int pivot = rd.nextInt(rt - lt + 1) + lt; // [lt,rt]
        int pivot_val = arr[pivot];
        int new_pivot = lt;

        swap(arr,pivot,rt);
        for(int i = lt;i<=rt;i++){
            if(arr[i] < pivot_val){
                swap(arr,i,new_pivot++);
            }
        }
        swap(arr,new_pivot,rt);
        return new_pivot;
    }

    /**
     * Quickselect
     * Search the smallest i that citation[i] >= N - i
     * @param citations
     * @return h index
     */   
    public int hIndex(int[] citations) {
        int lt = 0;
        int rt = citations.length - 1;
        int hIndex = 0;
        while (lt <= rt) {
            int pivot = partition(citations, lt, rt);
            if (citations[pivot] >= citations.length - pivot) {
                hIndex = citations.length - pivot;
                rt = pivot - 1;
            } else
                lt = pivot + 1;
        }
        return hIndex;
    }
}

File: 274 H-Index.py
'''
274. H-Index
https://leetcode.com/problems/h-index/

Given an array of citations (each citation is a non-negative integer) of a researcher,
write a function to compute the researcher's h-index.

According to the definition of h-index on Wikipedia:
"A scientist has index h if h of his/her N papers have at least h citations each,
and the other N − h papers have no more than h citations each."

Example:

Input: citations = [3,0,6,1,5]
Output: 3 
Explanation: [3,0,6,1,5] means the researcher has 5 papers in total and each of them had 
             received 3, 0, 6, 1, 5 citations respectively. 
             Since the researcher has 3 papers with at least 3 citations each and the remaining 
             two with no more than 3 citations each, her h-index is 3.
    
Note: If there are several possible values for h, the maximum one is taken as the h-index.
'''
class Solution:
    def hIndex(self, citations: List[int]) -> int:
        if not citations:
            return 0
        
        right = len(citations)
        left = 0
        hIndex = -1
        while left <= right:
            h = left + (right - left) // 2          
            moreThan = sum(c >= h for c in citations)
            lessThan = sum(c <= h for c in citations)
            
            '''
            This question description should be clearer.
            There can be more than h paper whose citation is at least h,
            similary there can be more than h paper whose citation is equal or less than h
            '''
            if moreThan >= h:
                if lessThan >= len(citations) - h:
                    hIndex = h
                left = h + 1
            else:
                right = h - 1
                
        return hIndex
